File: app/src/main/java/com/makingiants/todayhistory/utils/refresh_layout/CustomScrollSwipeRefreshLayout.java
package com.makingiants.todayhistory.utils.refresh_layout;

import android.content.Context;
import android.support.annotation.Nullable;
import android.support.v4.widget.SwipeRefreshLayout;
import android.util.AttributeSet;
import android.widget.ScrollView;

/**
 * Use {@link ScrollEnabler} to know when {@link CustomScrollSwipeRefreshLayout} can scroll up.
 * It's usefull when the mView have multiple inner views that are not inside a {@link ScrollView}
 */
public class CustomScrollSwipeRefreshLayout extends SwipeRefreshLayout {
  @Nullable private ScrollEnabler mScrollEnabler;

  public CustomScrollSwipeRefreshLayout(Context context, AttributeSet attrs) {
    super(context, attrs);
  }

  public void setScrollEnabler(@Nullable ScrollEnabler scrollEnabler) {
    mScrollEnabler = scrollEnabler;
  }

  @Override
  public boolean canChildScrollUp() {
    if (mScrollEnabler != null) {
      return mScrollEnabler.canScrollUp();
    }
    return super.canChildScrollUp();
  }
}

File: app/src/main/kotlin/com/makingiants/todayhistory/utils/refresh_layout/CustomScrollSwipeRefreshLayout.kt
package com.makingiants.todayhistory.utils.refresh_layout

import android.content.Context
import android.support.v4.widget.SwipeRefreshLayout
import android.util.AttributeSet

/**
 * Use [ScrollEnabler] to know when [CustomScrollSwipeRefreshLayout] can scroll up.
 * It's usefull when the mView have multiple inner views that are not inside a [ScrollView]
 */
class CustomScrollSwipeRefreshLayout(context: Context, attrs: AttributeSet) : SwipeRefreshLayout(context, attrs) {
    private var mScrollEnabler: ScrollEnabler? = null

    fun setScrollEnabler(scrollEnabler: ScrollEnabler?) {
        mScrollEnabler = scrollEnabler
    }

    override fun canChildScrollUp(): Boolean {
        if (mScrollEnabler != null) {
            return mScrollEnabler!!.canScrollUp()
        }
        return super.canChildScrollUp()
    }
}
